### PR TITLE
Flink: Move ParquetReader to LogicalTypeAnnotationVisitor

### PR DESCRIPTION
**Pull request by Fokko**
This will enable nanosecond timestamp support, since the OriginalType does not represent the nanosecond timestamp.

### CODE REVIEW

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java [hunk 1]
```diff
@@ -193,6 +195,122 @@ public ParquetValueReader<?> map(
...
+            return Optional.of(new MillisToTimestampReader(desc));
+          }
+        } else if (timestampLogicalType.getUnit() == LogicalTypeAnnotation.TimeUnit.MICROS) {
+          if (timestampLogicalType.isAdjustedToUTC()) {
+            return Optional.of(new MicrosToTimestampTzReader(desc));
+          } else {
+            return Optional.of(new MicrosToTimestampReader(desc));
+          }
+        }
+        return LogicalTypeAnnotation.LogicalTypeAnnotationVisitor.super.visit(timestampLogicalType);
```
Comment (pvary) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java:288]: Nit: newline

File: flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java [hunk 2]
```diff
@@ -193,6 +195,122 @@ public ParquetValueReader<?> map(
...
+        if (width <= 32) {
+          if (expected.typeId() == Types.LongType.get().typeId()) {
+            return Optional.of(new ParquetValueReaders.IntAsLongReader(desc));
+          } else {
+            return Optional.of(new ParquetValueReaders.UnboxedReader<>(desc));
+          }
+        } else if (width <= 64) {
+          return Optional.of(new ParquetValueReaders.UnboxedReader<>(desc));
+        }
+        return LogicalTypeAnnotation.LogicalTypeAnnotationVisitor.super.visit(intLogicalType);
```
Comment (pvary) [inline flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java:304]: Nit: newline

**Comment by pvary** [comment]
Could we add a test too? Minimally for the new timestamp?

**Comment by Fokko** [comment]
@pvary There are already quite a few tests: https://github.com/apache/iceberg/blob/main/flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkParquetReader.java

This writes and then reads a Parquet files using the visitor. I've added another test, but it is very hard to assert the readers since they are not public :(

File: gradle.properties [hunk 3]
```diff
@@ -16,7 +16,7 @@
 jmhOutputPath=build/reports/jmh/human-readable-output.txt
 jmhJsonOutputPath=build/reports/jmh/results.json
 jmhIncludeRegex=.*
-systemProp.defaultFlinkVersions=1.19
+systemProp.defaultFlinkVersions=1.18
```
Comment (pvary) [inline gradle.properties:19]: Fix this please

**Comment by pvary** [comment]
> @pvary There are already quite a few tests: https://github.com/apache/iceberg/blob/main/flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkParquetReader.java
> 
> This writes and then reads a Parquet files using the visitor. I've added another test, but it is very hard to assert the readers since they are not public :(

Does this test both the ns, and the ms timestamps?

File: gradle.properties (hunk 3 above)
Comment (Fokko) [reply gradle.properties:19]: Oops, thanks!
```suggestion
systemProp.defaultFlinkVersions=1.19
```

**Comment by Fokko** [comment]
> Does this test both the ns, and the ms timestamps?

No, not yet. I want to do that in a separate PR, but this change needs to be done before that. In Parquet, there is no `OriginalType` (Deprecated) for the Nanosecond timestamp, therefore we have to use the `LogicalTypeAnnotation` instead. Once https://github.com/apache/iceberg/pull/9008 has been merged, we can add support for Flink as well 👍 

https://github.com/apache/iceberg/pull/9719/files#diff-8d8e70413502d7efe586124264ffb87f49070dee35ac835a8ecdaac03df70196R273-R287

**Comment by pvary** [comment]
Thanks @Fokko for the PR!
Please do not forget the PR for the other Flink versions as well.

**Comment by Fokko** [comment]
Thanks @pvary for the review 👍

**Pull request merged**